core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/controller/Controller.scala

```diff
@@ -109,21 +111,34 @@ class Controller(
         )
       }.toSeq
 
+    // open all operators
+    val openOperatorRequests: Seq[Future[Unit]] =
+      workflow.getAllWorkers.map { workerID: ActorVirtualIdentity =>
+        asyncRPCClient.send(
+          OpenOperator(),
+          workerID
+        )
+      }.toSeq
+
     Future
       .collect(
-        initializeOperatorLogicRequests ++ activateLinkRequests
+        initializeOperatorLogicRequests
       )
-      .onSuccess({ _ =>
-        workflow.getAllOperators.foreach(_.setAllWorkerState(READY))
-        asyncRPCClient.sendToClient(WorkflowStatusUpdate(workflow.getWorkflowStatus))
-        context.become(running)
-        unstashAll()
-      })
       .onFailure((err: Throwable) => {
         logger.error("Failure when sending Python UDF code", err)
         // report error to frontend
         asyncRPCClient.sendToClient(FatalError(err))
       })
+      .flatMap(_ => Future.collect(activateLinkRequests))
+      .flatMap({ _ =>
+        Future {
+          workflow.getAllOperators.foreach(_.setAllWorkerState(READY))
+          asyncRPCClient.sendToClient(WorkflowStatusUpdate(workflow.getWorkflowStatus))
```

Comment:
we can move the above 2 lines to a new flatmap after "openOperatorRequests" so that the user sees "initializing" when the operator is opening

Reply:
moved after open operator requests.